libraries/standard/http/src/http_client.c

```diff
@@ -40,12 +116,128 @@ HTTPStatus_t HTTPClient_InitializeRequestHeaders( HTTPRequestHeaders_t * pReques
 /*-----------------------------------------------------------*/
 
 HTTPStatus_t HTTPClient_AddHeader( HTTPRequestHeaders_t * pRequestHeaders,
-                                   const char * pName,
-                                   size_t nameLen,
+                                   const char * pField,
+                                   size_t fieldLen,
                                    const char * pValue,
                                    size_t valueLen )
 {
-    return HTTP_NOT_SUPPORTED;
+    HTTPStatus_t returnStatus = HTTP_INTERNAL_ERROR;
+
+    /* Check for NULL parameters. */
+    if( pRequestHeaders == NULL )
+    {
+        IotLogError( "Parameter check failed: pRequestHeaders interface is NULL." );
+        returnStatus = HTTP_INVALID_PARAMETER;
+    }
+    else if( pRequestHeaders->pBuffer == NULL )
+    {
+        IotLogError( "Parameter check failed: pRequestHeaders->pBuffer is NULL." );
+        returnStatus = HTTP_INVALID_PARAMETER;
+    }
+    else if( ( pField == NULL ) )
+    {
+        IotLogError( "Parameter check failed: pField is NULL." );
+        returnStatus = HTTP_INVALID_PARAMETER;
+    }
+    else if( ( pValue == NULL ) )
+    {
+        IotLogError( "Parameter check failed: pValue is NULL." );
+        returnStatus = HTTP_INVALID_PARAMETER;
+    }
+    else
+    {
+        /* Empty else MISRA 15.7 */
+    }
+
+    /* Check if header field is long enough for length to overflow. */
+    if( fieldLen > ( UINT32_MAX >> 2 ) )
+    {
+        IotLogErrorWithArgs( "Parameter check failed: fieldLen must be less than %d.",
+                             ( UINT32_MAX >> 2 ) );
+        returnStatus = HTTP_INVALID_PARAMETER;
+    }
+    else if( fieldLen == 0U )
+    {
+        IotLogError( "Parameter check failed: fieldLen must be greater than 0." );
+        returnStatus = HTTP_INVALID_PARAMETER;
+    }
+    else
+    {
+        /* Empty else MISRA 15.7 */
+    }
+
+    /* Check if header value is long enough for length to overflow. */
+    if( valueLen > ( UINT32_MAX >> 2 ) )
+    {
+        IotLogErrorWithArgs( "Parameter check failed: valueLen must be less than %d.",
+                             ( UINT32_MAX >> 2 ) );
+        returnStatus = HTTP_INVALID_PARAMETER;
+    }
+    else if( valueLen == 0U )
+    {
+        IotLogError( "Parameter check failed: valueLen must be greater than 0." );
+        returnStatus = HTTP_INVALID_PARAMETER;
+    }
+    else
+    {
+        /* Empty else MISRA 15.7 */
+    }
+
+    if( returnStatus != HTTP_INVALID_PARAMETER )
+    {
+        /* "Content-Length" header must not be set by user if
```

Comment:
Why prevent the user from adding these headers? The users can very well skip the `InitializeRequestHeaders` call, and choose to add them through this function. 
Having this logic couples this function with knowledge of the `InitializeRequestHeader` function, which is not ideal.
Also, having the string comparison operations adds unnecessary performance cost every time a header is added through this function.

Reply:
You have a point there. Maybe @sarenameas can provide more insight.

Reply:
Yes let's go with archit's comment then we can remove the flags in the request headers.